REMOVE tslint, and use eslint for everything

ndelangen opened the pull request:
Issue: TSlint will get replaced with ESlint at some point (see [this article](https://medium.com/palantir/tslint-in-2019-1a144c2317a9) for explanations on why), having 2 linters is annoying:

- more processes to run
- more caching to configure
- more editor config

## What I did
- CHANGE the Eslint config so it can parse .ts,.tsx,.mjs files
- REMOVE tslint
- now that eslint rules are having an effect on typescript code, I did a ton of fixes
- had to do some additional fixes to avoid using ts-loader, related: https://github.com/storybooks/storybook/pull/6566

ndelangen committed:
REMOVE tslint, and use eslint for everything

ndelangen committed:
just a few little linting issues fixed

ndelangen committed:
FIX yarn.lock

ndelangen committed:
Merge branch 'next' into tech/remove-tslint

# Conflicts:
#	yarn.lock

ndelangen committed:
FIX snapshots

ndelangen committed:
REMOVE ---fix flag

shilman commented on addons/a11y/src/components/Report/HighlightToggle.tsx:
```diff
@@ -99,34 +103,55 @@ class HighlightToggle extends Component<ToggleProps> {
     }
   }
 
+  onToggle = (): void => {
```
What are these changes?

ndelangen replied:
Linting fixes, typescript files were missing a ton of useful eslint rules!

Now all eslint rules apply to typescript code as well 🎉

ndelangen committed:
REMOVE info about yarn linking && ADD info about linting

ndelangen committed:
Merge branch 'next' into tech/remove-tslint

ndelangen committed:
FIX yarn.lock

ndelangen committed:
Merge branch 'next' into tech/remove-tslint

# Conflicts:
#	yarn.lock

ndelangen committed:
FIX angular test & storybook build

ndelangen committed:
FIX lint issue

ndelangen committed:
CHANGE the loader for most code in the official example, unify js & ts loading

ndelangen committed:
FIX cache dir

shilman commented on examples/official-storybook/webpack.config.js:
```diff
@@ -5,7 +5,40 @@ module.exports = async ({ config }) => ({
+            loader: 'babel-loader',
+            options: {
+              cacheDirectory:
+                '/Users/dev/Projects/GitHub/storybook/core/examples/official-storybook/node_modules/.cache/storybook',
```
?!!!

gaetanmaisse replied:
Fixed in next commit :youhou:

shilman reviewed (changes requested):
Love it! One weird configuration change above, otherwise looks fantastic 🎉

gaetanmaisse commented on addons/a11y/src/components/ColorBlindness.tsx:
```diff
@@ -63,7 +63,8 @@ export class ColorBlindness extends Component<ColorBlindnessProps, ColorBlindnes
   onVisibilityChange = (s: boolean) => {
-    if (this.state.expanded !== s) {
+    const { expanded } = this.state;
+    if (expanded !== s) {
```
Which rule causes this? [destructuring-assignment](https://github.com/yannickcr/eslint-plugin-react/blob/master/docs/rules/destructuring-assignment.md) ?

ndelangen replied:
yeah, something react destructuring props & state

ndelangen replied:
I agree with this rule mostly, though I can be a bit annoying if you only have 1 prop or 1 state property.

gaetanmaisse commented on addons/contexts/src/shared/types.d.ts:
```diff
@@ -21,11 +22,11 @@ export declare interface AddonSetting {
     props: GenericProp;
     default?: boolean;
-  }>;
+  }[];
```
Not sure about this one, I think `Array<>` is easier to read. What other devs are thinking about it?

ndelangen replied:
I agree personally on that I find `Array<X>` to be easier to read then `X[]`, but this is a recommended rule from the typescript folk

gaetanmaisse commented on examples/cra-ts-kitchen-sink/src/react-app-env.d.ts:
```diff
@@ -1 +1 @@
-/// <reference types="react-scripts" />
+// / <reference types="react-scripts" />
```
⚠️ Nop, this is not a comment but a [triple slash directive](https://www.typescriptlang.org/docs/handbook/triple-slash-directives.html).

ndelangen committed:
FIX incorrect triple slash something

what a strange world we live in where a triple slash means something other then a code comment

shilman reviewed (approved):
Beautiful! 👌

Hypnosphi commented on .eslintrc.js:
```diff
@@ -94,13 +113,21 @@ module.exports = {
     ],
+    '@typescript-eslint/no-var-requires': ignore,
+    '@typescript-eslint/camelcase': ignore,
+    '@typescript-eslint/no-unused-vars': ignore,
```
Why?

Hypnosphi commented on examples/official-storybook/webpack.config.js:
```diff
@@ -5,7 +5,39 @@ module.exports = async ({ config }) => ({
+              presets: [
+                ['@babel/preset-env', { shippedProposals: true, useBuiltIns: 'usage', corejs: 2 }],
+                '@babel/preset-typescript',
+                ['babel-preset-minify', { builtIns: false, mangle: false }],
```
@ndelangen shouldn't we apply this one only in prod mode?

ndelangen replied:
yeah
